File: Algorithms/Swift/Tree/101_Symmetric_Tree.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.left = nil
 *         self.right = nil
 *     }
 * }
 */
class Solution {
    func isSymmetric(root: TreeNode?) -> Bool {
        if root == nil {
            return true
        }

        return helper(root!.left, root!.right)
    }

    private func helper(root1: TreeNode?, _ root2: TreeNode?) -> Bool {
        if root1 == nil && root2 == nil {
            return true
        }

        if root1 == nil && root2 != nil || root1 != nil && root2 == nil {
            return false
        }

        if root1!.val != root2!.val {
            return false
        }

        return helper(root1!.left, root2!.right) && helper(root1!.right, root2!.left)
    }
}

File: Algorithms/C++/Easy/101_Symmetric_Tree.cpp
// Recursion, O(n) Time, O(n) Space (Stack call)
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if (root == NULL) {
            return true;
        }

        return helper(root->left, root->right);
    }
private:
    bool helper(TreeNode* root1, TreeNode* root2) {
        if (root1 == NULL && root2 == NULL) {
            return true;
        }

        if (root1 == NULL || root2 == NULL) {
            return false;
        }

        if (root1->val != root2->val) {
            return false;
        }

        return helper(root1->left, root2->right) && helper(root1->right, root2->left);
    }
};

// Iterative, O(n) Time, O(n) Space
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        queue<TreeNode*> q;
        q.push(root);
        q.push(root);

        while (!q.empty()) {
            TreeNode* root1 = q.front();
            q.pop();

            TreeNode* root2 = q.front();
            q.pop();

            if (root1 == NULL && root2 == NULL) {
                continue;
            }
            if (root1 == NULL || root2 == NULL || root1->val != root2->val) {
                return false;
            }

            q.push(root1->left);
            q.push(root2->right);

            q.push(root1->right);
            q.push(root2->left);
        }

        return true;
    }
};
